UIU-2477: Unassign all permissions from a user with one click

OleksandrHladchenko1 opened the pull request:
UIU-2477: Unassign all permissions from a user with one click.
Refs: https://issues.folio.org/browse/UIU-2477.

https://user-images.githubusercontent.com/85172747/145831483-48344fe9-b375-4a40-ac1c-91b7a30857fb.mp4

zburke commented on src/components/PermissionsAccordion/PermissionsAccordion.js:
```diff
@@ -25,71 +28,51 @@ import PermissionModal from './components/PermissionsModal';
-  }
+  const [permissionModalOpen, setPermissionModalOpen] = useState(false);
+  const [unassignModalOpen, setUnassignModalOpen] = useState(false);
+  const [isUnassingButtonEnable, setIsUnassingButtonEnable] = useState(false);
```
Sorry; typo in these two method names: should be `...Unassign`.

zburke reviewed (commented):
Ah, crap. I just refactored a bunch of this myself in #1972. You can be forgiven for assuming I would never actually get to that work since I dragged it around for a half-dozen sprints 😬 . I love the feature that you've implemented though! I just wish I'd realized we were working in the same space before I hit "commit". 

LMK if you want me to pair up and work on a refactor together. Maybe it would go quicker that way since you know what needs to happen here, and I can explain the bits (actually, pretty small and well-isolated bits) that I changed.

OleksandrHladchenko1 commented:
> Ah, crap. I just refactored a bunch of this myself in #1972. You can be forgiven for assuming I would never actually get to that work since I dragged it around for a half-dozen sprints 😬 . I love the feature that you've implemented though! I just wish I'd realized we were working in the same space before I hit "commit".
> 
> LMK if you want me to pair up and work on a refactor together. Maybe it would go quicker that way since you know what needs to happen here, and I can explain the bits (actually, pretty small and well-isolated bits) that I changed.

Hi, @zburke! I think it's a good idea to pair up and work on a refactor together. I'm in.

OleksandrHladchenko1 commented:
Hi, @zburke! I refactored the code, you can look through my changes. But in `PermissionsAccordion.test.js` all tests are failed, and now I'm trying to fix it.

zburke commented:
@OleksandrHladchenko1, you need to mock `OnChange` from `react-final-form-listeners`. Otherwise, you're pullling in the real version, which desperately wants to be part of form and throws all these errors:
```
    useField must be used inside of a <Form> component
```
I pushed up a branch that separates the `<OnChange>` block into its own component so you could add coverage for that, too, if possible. Probably the `react-final-form-listeners` mock should be moved to `/test/jest/__mocks__`, otherwise you'll need to copy-past it into both the `PermissionsAccordion.test.js` and `EnableUnassignAll.test.js` files. Or not; NBD. But you'll need to beef it up a bit in order to be able to write a useful test for `EnableUnassignAll`.

HTH!

zburke commented on src/components/PermissionsAccordion/EnableUnassignAll.test.js:
```diff
@@ -0,0 +1,36 @@
+
+import EnableUnassignAll from './EnableUnassignAll';
+
+const onChengeMock = jest.fn();
```
whoops, typo: this should be `onChangeMock`.

zburke commented on test/jest/__mock__/reactFinalFormListeners.mock.js:
```diff
@@ -0,0 +1,9 @@
+  return {
+    ...jest.requireActual('react-final-form-listeners'),
+
+    OnChange: ({ children }) => <>{children()}</>,
```
You can mock the return value with `jest.fn` so lint doesn't whine about prop-types:
```
    OnChange: jest.fn(({ children }) => <>{children()}</>),
```

zburke reviewed (changes requested):
Sweet! The code looks great! Please just clean up the lint warning and typo and this is good to merge.

zburke reviewed (approved):
👏 👏 👏

zburke commented:
@OleksandrHladchenko1 , please use "Squash and merge" to keep the commit history tidy. Thanks for all your great work on this!

OleksandrHladchenko1 commented:
> @OleksandrHladchenko1 , please use "Squash and merge" to keep the commit history tidy. Thanks for all your great work on this!

@zburke, thank you for help!
